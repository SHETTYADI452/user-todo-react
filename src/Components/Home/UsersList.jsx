//Importing modules
import React from "react";

//exporting allusers and onclick to  index.jsx so that we can map the data of allUsers and perform a click event.
export const UsersList = ({ allUsers, onUserClick }) => {
  // will return a list of users
  return (
    <ul className="users-list-container">
      <ul>
        {allUsers?.map((user, index) => (
          <p
            key={`user-${index}`}
            className={`user-${index}`}
            id={`user-${index}`}
            onClick={() => onUserClick(user)}
          >
            {user.id}. {user.username}
          </p>
        ))}
      </ul>
    </ul>
  );
};
