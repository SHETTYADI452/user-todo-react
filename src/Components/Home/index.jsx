// importing modules
import React, { useEffect, useContext } from "react";
import { fetchUsers, setSelectedUser } from "../../actions/users";
import { UsersList } from "./UsersList";
import { UsersContext } from "../../Contexts/Users";
import { LoadingContext } from "../../Contexts/Loading";
import { useHistory } from "react-router-dom";
import ClipLoader from "react-spinners/ClipLoader";
import "../../styles/home.css";

//export default class
export default (props) => {
  const title = "USERS LIST";
  const history = useHistory();
  //Initializing users context
  const { state: userState, dispatch: userDispatch } = useContext(UsersContext);
  // Initializing loading context
  const { state: loadingState, dispatch: loadingDispatch } =
    useContext(LoadingContext);

  // with the help of useEffect Hook the as soons as the app renders the userdetails will be loaded with effect
  useEffect(() => {
    fetchUsers()(userDispatch, loadingDispatch);
  }, []);

  // This function will help to fetch user details with the help of diapatch and Navigate users to userdeatails page
  const handleOnUserClick = (user) => {
    // this will push to new page
    history.push(`/userdetail-todo/${user.id}`);
    // this function will invoke in action/user folder and set the userData
    setSelectedUser(user)(userDispatch, loadingDispatch);
  };

  return (
    <div>
      <center>
        <div id="page-title">
          <h1>{title}</h1>
        </div>

        <div className="userlist">
          {loadingState.isLoading ? (
            <div>
              <h3 className="loader">Fetching Users...</h3>
              <ClipLoader loading={loadingState.isLoading} size={150} />
            </div>
          ) : (
            <UsersList
              allUsers={userState.users}
              onUserClick={handleOnUserClick}
            />
          )}
        </div>
      </center>
    </div>
  );
};
