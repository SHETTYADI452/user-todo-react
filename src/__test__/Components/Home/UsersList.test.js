import React from "react";
import { mount } from "enzyme";
import { UsersList } from "../../../Components/Home/UsersList";

const mockHistoryPush = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}));

describe.only("Testing for the UsersList Component ", () => {
  let wrapper;
  const onUserClick = jest.fn();
  const allUsers = [
    {
      id: 1,
      username: "user 1",
    },
    {
      id: 2,
      username: "user 2",
    },
    {
      id: 3,
      username: "user 3",
    },
  ];

  const defaultProps = {
    allUsers,
    onUserClick,
  };

  beforeEach(() => {
    wrapper = mount(<UsersList {...defaultProps} />);
  });

  it("should match the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should check for the users list Container", () => {
    expect(wrapper.find(".users-list-container")).toBeTruthy();
    expect(wrapper.find(".users-list-container")).toHaveLength(1);
  });
  it("should check the mapping of the users list", () => {
    allUsers?.map((user, index) => {
      expect(wrapper.find(`#user-${index}`).length).toBe(1);
      expect(wrapper.find(`#user-${index}`).text()).toEqual(
        `${user.id}. ${user.username}`
      );
      wrapper.find(`#user-${index}`).simulate("click");

      const userLink = wrapper.find(`#user-${index}`);
      userLink.simulate("click");
      expect(onUserClick).toHaveBeenCalledWith(allUsers[0]);
    });
  });
});
