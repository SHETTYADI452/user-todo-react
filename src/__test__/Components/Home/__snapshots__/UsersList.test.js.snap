// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing for the UsersList Component  should match the snapshot 1`] = `ReactWrapper {}`;
